rep-2012.rst

```diff
@@ -0,0 +1,425 @@
+- Enable the sending of metadata and service contents
+
+By default, the event publishing feature is off for all clients and all services so users do not pay for a feature they do not plan to use.
+Furthermore, node authors may opt-in by default or disable the service introspection feature altogether as they see fit.
+
+Quality of Service
+------------------
+
+The service event topics proposed in this REP shall use the default quality of service settings [9]_.
+
+Security
+--------
+
+Enabling service introspection creates more attack surface for an existing ROS system by adding 2*N more topics (where N is the number of services with the feature enabled).
+These topics are vulnerable to undesired actors listening in on service communication or even interfering with parts of the system that may be relying on service events.
+
+Luckily, we can leverage the existing security feature for topics and services in ROS 2 (see SROS 2 [10]_).
+Any existing tooling for aiding users in setting up ROS security should consider the new service event topics introduced by this REP (e.g. NoDL [11]_).
+
+
+Rationale
+=========
+
+The following sections summarize *why* certain design decisions were made and some of the alternatives considered.
+
+Configuring service introspection through API calls
+---------------------------------------------------
+
+There are a few reasons to configure introspection through API calls.
+
+First, enabling or disabling introspection is fundamentally a per-client or per-service action.
+In most scenarios, users probably will not want to enable service introspection on all clients and services at once, as this will greatly increase network traffic.
+
+Next, it makes sense to have users of the rcl client/server API not have to specified arguments that will never be used if they don't enable introspection.
+By having a separate API for this, only API users concerned with enabling the introspection feature need to provide the feature.
+
+Finally, by having a separate API call for introspection, the API behavior ends up being completely orthogonal.
+That is, users can cycle between having introspection off, metadata only, or contents sent, and the system will do the correct thing.
+
+One downside of using APIs for configuration is that there is no obvious way to configure the introspection feature at runtime.
+However, it is easy to hook up the API call to a ROS parameter (say), and control it through that.
+If that turns out to be a popular feature, the implementation can be extended to automatically expose this per-service introspection as a parameter.
+
+Only supporting one service per name
+------------------------------------
+
+It is technically possible to create more than one service with the same name (though not recommended).
+However, this is generally not recommended and may be forbidden in the future.
+Therefore, as far as this REP is concerned, creating multiple services with the same name is undefined behavior.
```

Comment:
Would it make sense to mention this is undefined behavior in the specification section since it deals with behavior?